odoo/addons/base/tests/test_ir_cron.py

```diff
@@ -251,16 +223,117 @@ def test_cron_process_job(self):
 
+    def test_cron_retrigger(self):
+        Trigger = self.env['ir.cron.trigger']
+        Progress = self.env['ir.cron.progress']
+
+        def run(self):
+            count = Progress.search_count([('done', '=', 1)])
+            if count < 5:
+                return 1, 1
+            elif count < 10:
+                self.env['ir.cron']._notify_progress(1, 1)
+            else:
+                return 1, 0
+
+        self.cron._trigger()
+        self.env.flush_all()
+        self.registry.enter_test_mode(self.cr)
+        try:
+            with patch.object(self.registry['ir.actions.server'], 'run', run):
+                self.registry['ir.cron']._process_job(
+                    self.registry.db_name,
+                    self.registry.cursor(),
+                    self.cron.read(load=None)[0]
+                )
+        finally:
+            self.registry.leave_test_mode()
+
+        self.assertEqual(
+            Progress.search_count([('done', '=', 1), ('cron_id', '=', self.cron.id)]), 10,
+            'There should be 10 progress log for this cron',
+        )
+        self.assertEqual(
+            Trigger.search_count([('cron_id', '=', self.cron.id)]), 1,
+            "One trigger should have been kept",
+        )
+
+        self.env.flush_all()
+        self.registry.enter_test_mode(self.cr)
+        try:
+            with patch.object(self.registry['ir.actions.server'], 'run', run):
+                self.registry['ir.cron']._process_job(
+                    self.registry.db_name,
+                    self.registry.cursor(),
+                    self.cron.read(load=None)[0]
+                )
+        finally:
+            self.registry.leave_test_mode()
+
+        self.assertEqual(
+            Trigger.search_count([('cron_id', '=', self.cron.id)]), 0,
+            "No trigger should have been kept",
+        )
+        self.assertEqual(
+            Progress.search_count([('done', '=', 1), ('cron_id', '=', self.cron.id)]), 11,
+            'There should be 11 progress log for this cron',
+        )
+
+    def test_cron_failed_increase(self):
+        self.cron._trigger()
+        self.env.flush_all()
+        self.registry.enter_test_mode(self.cr)
+        try:
+            with (
+                patch.object(self.registry['ir.cron'], '_callback', side_effect=Exception),
+                patch.object(self.registry['ir.cron'], '_notify_admin') as notify,
+            ):
+                self.registry['ir.cron']._process_job(
+                    self.registry.db_name,
+                    self.registry.cursor(),
+                    self.cron.read(load=None)[0]
+                )
+        finally:
+            self.registry.leave_test_mode()
+
+        self.env.invalidate_all()
+        self.assertEqual(self.cron.failures, 1, 'The cron should have failed once')
+        self.assertEqual(self.cron.active, True, 'The cron should still be active')
+        self.assertFalse(notify.called)
+
+        self.cron.failures = 5
```

Comment:
Before the cron's fifth failure, the failure count should be 4:
```suggestion
        self.cron.failures = 4
```